### ISSUE TITLE
Stir single gene pool instead of shuffling multiple pools

### ISSUE THREAD

**Issue by MarkZH** [MarkZH/Genetic_Chess#84]
Instead of individual gene pools, have one large pool and stir the order to create match-ups. By "stir" I mean a shuffle where most elements end up near where they started. This is controlled by a parameter `k` in [0, 1] where larger values result in larger average movement. This `k` would be read from the gene pool configuration file (or maybe an "average movement distance" parameter).

The code would look something like:
```cpp
template<typename Indexable_Container>
Random::stir(double k, Indexable_Container& list)
{
    auto gen = std::mt19937_64(std::random_device());
    auto distr = std::piecewise_constant_distribution<double>(list.size(),
                                                              0,
                                                              list.size() - 1,
                                                              [k](auto i) { return std::pow(k, i); });
    for(int i = 0; i < pool.size() - 1; ++i)
    {
        j = size_t(distr(gen)) + i;
        std::swap(listi], list[j]);
    }
}
```
When `k == 0` there is no movement and when `k == 1` it is equivalent to a standard shuffle with all orderings equally likely. When `k > 1`, I imagine that would result in stirring the list with some other value `k2 < 1` and then reversing the list.

The smaller the value of `k`, the slower the AIs wander about when pairing up for games/mating and vice versa. This keeps unrelated AIs far apart for a long time to preserve diversity. This simulates a landscape where the AIs have to wander about to find opponents/mates and `k` indicates how fast they travel.

References:
https://en.cppreference.com/w/cpp/numeric/random/piecewise_constant_distribution

**Comment by MarkZH** [comment]
The specific algorithm above doesn't work (first, `std::discrete_distribution` takes `double`s for weights, so jerry-rigging `std::piecewise_constant_distribution` was unnecessary). This is mainly because stirrings where the end of the list gets more stirred than the beginning are more likely. The exponential probability function was probably the culprit, and I don't know if I could figure out another one that works. The below code is much more unbiased:

```cpp
template<typename Ordered_Iterator>
void stir(double std_dev, Ordered_Iterator begin, Ordered_Iterator end)
{
    using value_type = std::remove_reference_t<decltype(*begin)>;

    static auto gen = std::mt19937_64(std::random_device{}());
    auto distr = std::normal_distribution<double>(0.0, std_dev);

    auto positions = std::map<double, value_type>{};
    for(auto it = begin; it != end; ++it)
    {
        positions[positions.size() + distr(gen)] = *it;
    }
    std::transform(positions.begin(), positions.end(), begin, [](const auto& p) { return p.second; });
}
```
In other words, take each index as a 1-D position, add a random normally (or laplacianly) distributed value, and order by the new position. The first and last positions tend to stay put, which may or may not be a problem. Maybe I could bring back the reflective_clamp function.

**Comment by MarkZH** [comment]
Instead of reflectively clamping, make the world round: `while(position < 0.0) position += pool.size();`.
